File: validateai/supabase/functions/_shared/opsAlert.ts
// Alerting de operaciones para las Edge Functions.
//
// Port del helper de mp-sync (services/mercado-publico/src/infrastructure/
// ops-alert). Mismos canales y mismo formato de embed, para que la sala de
// control se lea igual venga de donde venga el aviso.
//
// POR QUÉ ESTE ARCHIVO EXISTE
// ---------------------------
// Hasta ahora sólo la ingesta de Mercado Público hablaba a Discord. El motor de
// IA y el gateway público —donde vive el producto que se le cobra a alguien—
// eran completamente mudos: si `ai-validate` degradaba o `api-v1` empezaba a
// devolver 500, no había ninguna señal fuera de los logs.
//
// Los canales se separan por QUÉ HACER al ver el mensaje:
//   incidentes   → algo está roto AHORA. Sólo rojo.
//   latido       → corridas programadas que terminaron bien.
//   degradacion  → lo que "funciona" mientras miente (RAG sin corpus, mocks,
//                  fallbacks sirviendo en lugar del dato real).
//   negocio      → señal de producto: cuotas agotadas, tiers, leads.
//
// Nunca lanza y nunca bloquea: un fallo de alerting no debe romper el flujo que
// lo emitió. Todos los llamadores deben usar `void sendOpsAlert(...)`.

export type OpsLevel = 'info' | 'warn' | 'error';
export type OpsChannel = 'incidentes' | 'latido' | 'degradacion' | 'negocio';

export interface OpsField {
  name: string;
  value: string;
  inline?: boolean;
}

export interface OpsAlert {
  level: OpsLevel;
  title: string;
  detail?: string;
  fields?: OpsField[];
  /** Contexto de origen al pie. Por convención: el nombre de la función. */
  footer?: string;
  channel?: OpsChannel;
  dedupeKey?: string;
}

const EMOJI: Record<OpsLevel, string> = { error: '🔴', warn: '🟡', info: '🟢' };

/** Color de la barra lateral: la señal que se lee sin leer. */
const COLOR: Record<OpsLevel, number> = {
  error: 0xe04f5f,
  warn: 0xe0a44f,
  info: 0x4fe08a,
};

const NOMBRE_CANAL: Record<OpsChannel, string> = {
  incidentes: 'Incidentes',
  latido: 'Latido',
  degradacion: 'Degradación',
  negocio: 'Negocio',
};

function urlDeCanal(channel: OpsChannel): string | undefined {
  const incidentes = Deno.env.get('OPS_WEBHOOK_URL');
  const porCanal: Record<OpsChannel, string | undefined> = {
    incidentes,
    latido: Deno.env.get('OPS_WEBHOOK_LATIDO'),
    degradacion: Deno.env.get('OPS_WEBHOOK_DEGRADACION'),
    negocio: Deno.env.get('OPS_WEBHOOK_NEGOCIO'),
  };
  // Sin canal propio cae a incidentes: preferible un canal mezclado a un aviso
  // mudo.
  return porCanal[channel] ?? incidentes;
}

/**
 * Dedupe en memoria del isolate. Las Edge Functions se reciclan seguido, así
 * que esto NO es un dedupe distribuido: sólo corta la repetición dentro de una
 * misma invocación o de invocaciones cercanas en el mismo isolate. Alcanza para
 * el caso real (el mismo aviso disparándose en un bucle) sin infraestructura.
 */
const DEDUPE_MS = 30 * 60 * 1000;
const ultimoEnvio = new Map<string, number>();

function debeEnviar(key: string): boolean {
  const ahora = Date.now();
  const previo = ultimoEnvio.get(key);
  if (previo != null && ahora - previo < DEDUPE_MS) return false;
  ultimoEnvio.set(key, ahora);
  if (ultimoEnvio.size > 100) {
    for (const [k, t] of ultimoEnvio) {
      if (ahora - t >= DEDUPE_MS) ultimoEnvio.delete(k);
    }
  }
  return true;
}

function textoPlano(a: OpsAlert): string {
  const partes = [`${EMOJI[a.level]} **${a.title}**`];
  if (a.detail) partes.push(a.detail);
  if (a.fields?.length) partes.push(a.fields.map((f) => `${f.name}: ${f.value}`).join(' · '));
  return partes.join('\n');
}

export async function sendOpsAlert(alert: OpsAlert): Promise<void> {
  const channel = alert.channel ?? (alert.level === 'error' ? 'incidentes' : 'latido');

  // Siempre al log, haya webhook o no.
  const linea = `[ops-alert:${channel}] ${alert.title}${alert.detail ? ` — ${alert.detail}` : ''}`;
  if (alert.level === 'error') console.error(linea);
  else console.warn(linea);

  const url = urlDeCanal(channel);
  if (!url) return;
  if (!debeEnviar(`${channel}:${alert.dedupeKey ?? alert.title}`)) return;

  const plano = textoPlano(alert);
  const embed = {
    title: `${EMOJI[alert.level]}  ${alert.title}`.slice(0, 256),
    ...(alert.detail ? { description: alert.detail.slice(0, 4096) } : {}),
    color: COLOR[alert.level],
    ...(alert.fields?.length
      ? {
          fields: alert.fields.slice(0, 25).map((f) => ({
            name: f.name.slice(0, 256),
            value: (f.value || '—').slice(0, 1024),
            inline: f.inline ?? true,
          })),
        }
      : {}),
    footer: {
      text: `${alert.footer ?? 'edge'} · ${NOMBRE_CANAL[channel]}`.slice(0, 2048),
    },
    timestamp: new Date().toISOString(),
  };

  try {
    const res = await fetch(url, {
      method: 'POST',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify({ embeds: [embed], text: plano }),
    });
    if (!res.ok) {
      // Un embed mal formado da 400 y el aviso se perdería. Mejor feo que mudo.
      await fetch(url, {
        method: 'POST',
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify({ content: plano, text: plano }),
      }).catch(() => {});
    }
  } catch (err) {
    console.warn('[ops-alert] fallo al enviar webhook:', err);
  }
}

File: validateai-financial-worker/api/ops_alert.py
"""
api/ops_alert.py — Alerting de operaciones hacia la sala de control en Discord.

Port del helper que usan mp-sync (TypeScript) y las Edge Functions (Deno).
Mismos canales, mismos colores y mismo formato de embed, para que la sala se lea
igual venga de donde venga el aviso.

POR QUÉ EXISTE
--------------
BralidusPY era el último servicio mudo del ecosistema. Corre 9 jobs de ingesta
(FRED, CMF, BCCh, SEIA, concursal, empleo, radar…) disparados por cron externo:
si uno se caía, no había señal fuera de los logs de Railway. Es exactamente la
situación que dejó la ingesta de Mercado Público tres días detenida sin que
nadie lo notara.

Los canales se separan por QUÉ HACER al ver el mensaje:
    incidentes   → algo está roto AHORA. Sólo rojo.
    latido       → corridas programadas que terminaron. El valor está en el
                   hueco: si dejan de llegar, algo se detuvo.
    degradacion  → lo que "funciona" mientras miente.

Nunca lanza: un fallo de alerting jamás debe romper el flujo que lo emitió.
Todas las funciones de acá capturan sus propias excepciones.
"""

from __future__ import annotations

import logging
import os
import threading
import time
from typing import Literal

import requests

log = logging.getLogger(__name__)

Nivel = Literal["info", "warn", "error"]
Canal = Literal["incidentes", "latido", "degradacion"]

_EMOJI: dict[str, str] = {"error": "🔴", "warn": "🟡", "info": "🟢"}

# Color de la barra lateral del embed: la señal que se lee sin leer.
_COLOR: dict[str, int] = {"error": 0xE04F5F, "warn": 0xE0A44F, "info": 0x4FE08A}

_NOMBRE_CANAL: dict[str, str] = {
    "incidentes": "Incidentes",
    "latido": "Latido",
    "degradacion": "Degradación",
}

# Ventana de dedupe: como máximo un aviso por clave en este lapso.
_DEDUPE_S = 30 * 60
_ultimo_envio: dict[str, float] = {}
_lock = threading.Lock()

_TIMEOUT_S = 8


def _url_de_canal(canal: str) -> str | None:
    """URL del canal, con caída a incidentes si ese canal no está configurado.

    Preferible un canal mezclado a un aviso mudo.
    """
    incidentes = os.getenv("OPS_WEBHOOK_URL")
    por_canal = {
        "incidentes": incidentes,
        "latido": os.getenv("OPS_WEBHOOK_LATIDO"),
        "degradacion": os.getenv("OPS_WEBHOOK_DEGRADACION"),
    }
    return por_canal.get(canal) or incidentes


def _debe_enviar(clave: str) -> bool:
    ahora = time.time()
    with _lock:
        previo = _ultimo_envio.get(clave)
        if previo is not None and ahora - previo < _DEDUPE_S:
            return False
        _ultimo_envio[clave] = ahora
        if len(_ultimo_envio) > 200:
            for k, t in list(_ultimo_envio.items()):
                if ahora - t >= _DEDUPE_S:
                    del _ultimo_envio[k]
    return True


def _texto_plano(nivel: str, titulo: str, detalle: str | None, campos: list[tuple[str, str]]) -> str:
    partes = [f"{_EMOJI[nivel]} **{titulo}**"]
    if detalle:
        partes.append(detalle)
    if campos:
        partes.append(" · ".join(f"{n}: {v}" for n, v in campos))
    return "\n".join(partes)


def send_ops_alert(
    *,
    nivel: Nivel,
    titulo: str,
    detalle: str | None = None,
    campos: list[tuple[str, str]] | None = None,
    footer: str | None = None,
    canal: Canal | None = None,
    dedupe_key: str | None = None,
) -> None:
    """Publica un aviso. Fire-and-forget: nunca lanza y nunca bloquea el flujo.

    `campos` son pares (nombre, valor) que se renderizan como columnas del embed.
    """
    campos = campos or []
    canal_final: str = canal or ("incidentes" if nivel == "error" else "latido")

    # Siempre al log, haya webhook o no.
    linea = f"[ops-alert:{canal_final}] {titulo}" + (f" — {detalle}" if detalle else "")
    if nivel == "error":
        log.error(linea)
    else:
        log.warning(linea)

    url = _url_de_canal(canal_final)
    if not url:
        return
    if not _debe_enviar(f"{canal_final}:{dedupe_key or titulo}"):
        return

    plano = _texto_plano(nivel, titulo, detalle, campos)
    embed: dict = {
        "title": f"{_EMOJI[nivel]}  {titulo}"[:256],
        "color": _COLOR[nivel],
        "footer": {"text": f"{footer or 'bralidus'} · {_NOMBRE_CANAL[canal_final]}"[:2048]},
        "timestamp": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
    }
    if detalle:
        embed["description"] = detalle[:4096]
    if campos:
        # Discord admite hasta 25 campos por embed.
        embed["fields"] = [
            {"name": n[:256], "value": (v or "—")[:1024], "inline": True} for n, v in campos[:25]
        ]

    try:
        res = requests.post(
            url,
            json={"embeds": [embed], "text": plano},
            timeout=_TIMEOUT_S,
        )
        if res.status_code >= 400:
            # Un embed mal formado da 400 y el aviso se perdería del todo.
            # Mejor feo que mudo.
            requests.post(url, json={"content": plano, "text": plano}, timeout=_TIMEOUT_S)
    except Exception as err:  # noqa: BLE001 — el alerting nunca debe propagar
        log.warning("[ops-alert] fallo al enviar webhook: %s", err)
